### PR TITLE
DBZ-7164 Support offset commit for the signal topic

### PR DESCRIPTION
**Pull request by vsantona**
Documentation: https://github.com/debezium/debezium/pull/5116

### CODE REVIEW

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/SignalsIT.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                .with(MySqlConnectorConfig.INCLUDE_SCHEMA_CHANGES, false)
+                .with(MySqlConnectorConfig.SIGNAL_DATA_COLLECTION, DATABASE.qualifiedTableName("debezium_signal"))
+                .with(CommonConnectorConfig.SIGNAL_POLL_INTERVAL_MS, 1)
+                .with(MySqlConnectorConfig.INCREMENTAL_SNAPSHOT_CHUNK_SIZE, 10)
+                .with(MySqlConnectorConfig.INCREMENTAL_SNAPSHOT_ALLOW_SCHEMA_CHANGES, true)
+                .with(CommonConnectorConfig.SCHEMA_NAME_ADJUSTMENT_MODE, CommonConnectorConfig.SchemaNameAdjustmentMode.AVRO);
+    }
+
+    @Test
+    public void givenOffsetCommitDisabledAndASignalSentWithConnectorRunning_whenConnectorComesBackUp_thenAllSignalsAreCorrectlyProcessed()
```
Comment (mfvitale) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/SignalsIT.java:102]: I would like to see also a test with the old behavior: signal sent during shutdown will not be processed on startup.

File: debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java [hunk 2]
```diff
@@ -163,15 +172,20 @@ public void reset(Object reference) {
...
                 .withDefault(ConsumerConfig.GROUP_ID_CONFIG, signalConfig.getString(GROUP_ID))
                 .withDefault(ConsumerConfig.FETCH_MIN_BYTES_CONFIG, 1) // get even the smallest message
                 .withDefault(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
                 .withDefault(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, 10000) // readjusted since 0.10.1.0
                 .withDefault(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class)
-                .withDefault(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class)
+                .withDefault(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        if (signalConfig.getBoolean(SIGNAL_CONSUMER_OFFSET_COMMIT_ENABLED)) {
+            confBuilder.withDefault(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, true)
+                    .withDefault(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
```
Comment (mfvitale) [inline debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: Is this enough? Should this act only when there is no initial offset in Kafka or if an offset is out of range? What about the restart after some processed records? Should be an explicit commit or auto commit present?

**Comment by mfvitale** [comment]
@vsantona thanks! I leaved couple of comments.

File: debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java (hunk 2 above)
Comment (mfvitale) [reply debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: Sorry, just saw the auto commit setting. Don't mind mu comment.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/SignalsIT.java (hunk 1 above)
Comment (vsantona) [reply debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/SignalsIT.java:102]: done

**Review by mfvitale — APPROVED** [review]

**Comment by mfvitale** [comment]
@vsantona Applied, thanks a lot!

File: debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java (hunk 2 above)
Comment (rk3rn3r) [inline debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: I randomly hit this PR and I was interested why you picked "earliest" here? The most probably case for not having an offset is either on a newly created connector or `kafka.group.id` or a purposely renamed connector `kafka.group.id` or consumer offsets . That might also be re-using an previously used group id accidentally.
Should in this case really all signals that existed before be re-processed?
(the code now looks a little different, but the property is still set to `earliest`)

Comment (mfvitale) [reply debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: You can read the reason behind this in [DBZ-7164](https://issues.redhat.com/browse/DBZ-7164)

**TL;DR**

The main problem was that when a connector is stopped if you send signals during this period them was not processed once the connector starts this because there was neither autocommit nor commit. 

Now the code is different since we have the auto-commit enabled by default and we just fallback to `earliest` in case of no offset.

Comment (rk3rn3r) [reply debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: Thank you. Yeah, that's a good explanation about `what` happened and I saw it from the ticket and the conversation on chat.
But let me ask again the last question from above: should in this case really all signals that existed before being re-processed aka `earliest`? This is also the case when starting a new "instance" of the kafka signal channel by renaming the consumer group, etc. I get that this can be (experienced) user's intention, but this can also be a source of error for less experienced users. The bus factor of `latest` is much lower. Maybe this could be configurable option and defaults to `latest`? Wdyt?
FYI @jpechane @Naros

Comment (Naros) [reply debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: But @rk3rn3r, if you deploy a random sink connector in Kafka Connect, when there are no offsets, that connector defaults to `earliest` unless you explicitly override the default behavior. So I would say what we are doing maps directly to the expected behavior for any other sink deployment. The expectation is that if the user needs a different strategy, they must override that.

Comment (rk3rn3r) [reply debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: That's not fully accurate. A Kafka consumer group (which is used for that behavior in source connectors) defaults to that. And for source connectors it makes sense.

I think signals have a different idea on how they work: you might not want to consume all signals every created (that are probably already tombstoned later in a stream and not yet removed, for example), but you normally want to read all data of a database (respecting several filter options etc).

And my take was not about the technical process, but the significant higher bus factor when `earliest` was used, but not meant. There is only a way with additional kafka property overwriting for the consumer config options to switch to `latest`. A dbz config option would make this more visibile.

Comment (mfvitale) [reply debezium-core/src/main/java/io/debezium/pipeline/signal/channels/KafkaSignalChannel.java:186]: @rk3rn3r If you still have some doubt I suggest to open a discussion on zulip commenting the current code. Since the code in this PR doesn't fully reflect the current code.
